modules/core/src/parallel.cpp

```diff
@@ -775,26 +796,86 @@ static inline int getNumberOfCPUsImpl()
       }
 
    }
-   return cpusAvailable ? cpusAvailable : 1;
+   return cpusAvailable;
+}
+#endif
+
+#if defined CV_HAVE_CGROUP
+static inline unsigned getNumberOfCPUsCFS(void)
+{
+    std::string file_contents = getFileContents("/sys/fs/cgroup/cpu/cpu.cfs_quota_us");
+    if(file_contents.empty())
+        return 0;
+
+    std::stringstream ss_period(file_contents);
+    int cfs_quota;
+    ss_period >> cfs_quota;
+
+    if(cfs_quota < 1) /* cfs_quota must not be 0 or negative */
+        return 0;
+
+    file_contents = getFileContents("/sys/fs/cgroup/cpu/cpu.cfs_period_us");
+    if(file_contents.empty())
+        return 0;
+
+    std::stringstream ss_quota(file_contents);
+    int cfs_period;
+    ss_quota >> cfs_period;
+
+    if(cfs_period < 1)
+        return 0;
+
+    return (unsigned) cfs_quota/cfs_period;
 }
 #endif
 
 int cv::getNumberOfCPUs(void)
 {
+#if defined CV_CXX11
+    /* Check for this standard C++11 way, we do not return directly because
+       running in a docker or K8s environment will mean this is the host
+       machines config not the containers or pods and as per docs this value
+       must be "considered only a hint" */
+    static unsigned concurentThreadsSupported = std::thread::hardware_concurrency(); /* If the value is not well defined or not computable, returns 0 */
+#else
+    static unsigned concurentThreadsSupported = 0; /* 0 means we have to find out some other way */
+#endif
+
 #if defined _WIN32
     SYSTEM_INFO sysinfo;
 #if (defined(_M_ARM) || defined(_M_ARM64) || defined(_M_X64) || defined(WINRT)) && _WIN32_WINNT >= 0x501
     GetNativeSystemInfo( &sysinfo );
 #else
     GetSystemInfo( &sysinfo );
 #endif
+    unsigned ncpus = min_non_zero(concurentThreadsSupported, sysinfo.dwNumberOfProcessors);
+    return ncpus != 0 ? ncpus : 1;
+#elif defined CV_HAVE_CGROUP && !defined __ANDROID__
+    unsigned ncpus = concurentThreadsSupported;
+    static unsigned ncpus_impl = (unsigned)getNumberOfCPUsImpl("/sys/fs/cgroup/cpuset/cpuset.cpus");
+    ncpus = min_non_zero(ncpus, ncpus_impl);
+
+    static unsigned ncpus_impl_cfs = getNumberOfCPUsCFS();
+    ncpus = min_non_zero(ncpus, ncpus_impl_cfs);
+
+    #if defined _GNU_SOURCE
+        cpu_set_t cpu_set;
+        if(!sched_getaffinity(0, sizeof(cpu_set), &cpu_set)) {
+            unsigned cpu_count_cpu_set = CPU_COUNT(&cpu_set);
+            ncpus = min_non_zero(ncpus, cpu_count_cpu_set);
+        }
+    #endif
+
+    static unsigned cpu_count_sysconf = (unsigned)sysconf( _SC_NPROCESSORS_ONLN );
+    ncpus = min_non_zero(ncpus, cpu_count_sysconf);
```

Comment:
Just split this block on 4 cases through sequence (moving "return" to the end would allow that).

`sched_getaffinity()` is applicable on Android too.

Reply:
@alalek, can I check cgroups for android as well?